pulp_file/tests/functional/conftest.py

```diff
@@ -192,3 +204,59 @@ def _file_fixture_gen_file_repo(**kwargs):
         return gen_object_with_cleanup(file_repo_api_client, kwargs)
 
     yield _file_fixture_gen_file_repo
+
+
+@pytest.fixture
+def gen_bad_response_fixture_server(gen_threaded_aiohttp_server):
+    """
+    This server will perform 3 bad responses for each file requested.
+
+    1st response will be incomplete, sending only half of the data.
+    2nd response will have corrupted data, with one byte changed.
+    3rd response will return error 429.
+    4th response will be correct.
+    """
+
+    def _gen_fixture_server(fixtures_root, ssl_ctx):
+        record = []
+        num_requests = defaultdict(int)
+
+        async def handler(request):
+            nonlocal num_requests
+            record.append(request)
+            relative_path = request.raw_path[1:]  # Strip off leading "/"
+            file_path = Path(fixtures_root) / Path(relative_path)
+            # Max retries is 3. So on fourth request, send full data
+            num_requests[relative_path] += 1
+            if "PULP_MANIFEST" in relative_path or num_requests[relative_path] % 4 == 0:
+                return web.FileResponse(file_path)
+
+            # On third request send 429 error, TooManyRequests
+            if num_requests[relative_path] == 3:
+                raise web.HTTPTooManyRequests
+
+            size = file_path.stat().st_size
+            response = web.StreamResponse(headers={"content-length": f"{size}"})
+            await response.prepare(request)
+            async with aiofiles.open(file_path, "rb") as f:
+                # Send only partial content causing aiohttp.ClientPayloadError if request num == 1
+                chunk = await f.read(size // 2)
+                await response.write(chunk)
+                # Send last chunk with modified last byte if request num == 2
+                if num_requests[relative_path] == 2:
+                    chunk2 = await f.read()
+                    await response.write(chunk2[:-1])
+                    await response.write(bytes([chunk2[-1] + 1]))
```

Comment:
Maybe just flip a bit to be safe:
```suggestion
                    await response.write(bytes([chunk2[-1] ^ 1]))
```